src/gazebo_mavlink_interface.cpp

```diff
@@ -538,192 +600,238 @@ void GazeboMavlinkInterface::OnUpdate(const common::UpdateInfo&  /*_info*/) {
 
 void GazeboMavlinkInterface::send_mavlink_message(const mavlink_message_t *message, const int destination_port)
 {
-  uint8_t buffer[MAVLINK_MAX_PACKET_LEN];
-  int packetlen = mavlink_msg_to_send_buffer(buffer, message);
-
-  struct sockaddr_in dest_addr;
-  memcpy(&dest_addr, &_srcaddr, sizeof(_srcaddr));
-
-  if (destination_port != 0) {
-    dest_addr.sin_port = htons(destination_port);
+    
+  if(serial_enabled_ && destination_port == 0) {
+    assert(message != nullptr);
+    if (!is_open()) {
+      gzerr << "Serial port closed! \n";
+      return;
+    }
+    
+    {
+      lock_guard lock(mutex);
+      
+      if (tx_q.size() >= MAX_TXQ_SIZE) {
+//         gzwarn << "TX queue overflow. \n";  
+      }
+      tx_q.emplace_back(message);
+    }
+    io_service.post(std::bind(&GazeboMavlinkInterface::do_write, this, true));
   }
+  
+  else {
+    uint8_t buffer[MAVLINK_MAX_PACKET_LEN];
+    int packetlen = mavlink_msg_to_send_buffer(buffer, message);
 
-  ssize_t len = sendto(_fd, buffer, packetlen, 0, (struct sockaddr *)&_srcaddr, sizeof(_srcaddr));
+    struct sockaddr_in dest_addr;
+    memcpy(&dest_addr, &_srcaddr, sizeof(_srcaddr));
 
-  if (len <= 0) {
-    printf("Failed sending mavlink message\n");
+    if (destination_port != 0) {
+      dest_addr.sin_port = htons(destination_port);
+    }
+ 
+    ssize_t len = sendto(_fd, buffer, packetlen, 0, (struct sockaddr *)&_srcaddr, sizeof(_srcaddr));
+
+    if (len <= 0) {
+      printf("Failed sending mavlink message\n");
+    }
   }
+  
 }
 
 void GazeboMavlinkInterface::ImuCallback(ImuPtr& imu_message) {
-  // frames
-  // g - gazebo (ENU), east, north, up
-  // r - rotors imu frame (FLU), forward, left, up
-  // b - px4 (FRD) forward, right down
-  // n - px4 (NED) north, east, down
-  math::Quaternion q_gr = math::Quaternion(
+  common::Time current_time = world_->GetSimTime();  
+  double dt = (current_time - last_imu_time_).Double();
+  if (dt >= imu_rate_)
```

Comment:
@arwagoner the imu rate can only influence the publishing of `HIL_SENSOR`, not `HIL_STATE_QUATERNION`. So what you find in https://github.com/PX4/sitl_gazebo/pull/157/files#diff-af95b4649436828244349b1766ab5c46R833 should be put in https://github.com/PX4/sitl_gazebo/pull/157/files#diff-af95b4649436828244349b1766ab5c46R789